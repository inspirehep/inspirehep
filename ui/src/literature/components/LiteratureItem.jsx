import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { LoginOutlined, FilePdfOutlined } from '@ant-design/icons';

import ArxivEprintList from './ArxivEprintList';
import LiteratureDate from './LiteratureDate';
import AuthorsAndCollaborations from '../../common/components/AuthorsAndCollaborations';
import PublicationInfoList from '../../common/components/PublicationInfoList';
import UrlsAction from './UrlsAction';
import DOILinkAction from './DOILinkAction';
import IconText from '../../common/components/IconText';
import ListItemAction from '../../common/components/ListItemAction';
import EditRecordAction from '../../common/components/EditRecordAction';
import ResultItem from '../../common/components/ResultItem';
import { LITERATURE } from '../../common/routes';
import EventTracker from '../../common/components/EventTracker';
import LiteratureTitle from '../../common/components/LiteratureTitle';
import ResponsiveView from '../../common/components/ResponsiveView';
import CiteModalActionContainer from '../containers/CiteModalActionContainer';
import ConferenceInfoList from './ConferenceInfoList';
import pluralizeUnlessSingle from '../../common/utils';
import {
  InlineUL,
  SEPARATOR_MIDDLEDOT,
} from '../../common/components/InlineList';

class LiteratureItem extends Component {
  render() {
    const { metadata, searchRank } = this.props;

    const title = metadata.getIn(['titles', 0]);
    const authors = metadata.get('authors');

    const fullTextLinks = metadata.get('fulltext_links');
    const dois = metadata.get('dois');
    const recordId = metadata.get('control_number');
    const citationCount = metadata.get('citation_count', 0);
    const authorCount = metadata.get('number_of_authors');
    const conferenceInfo = metadata.get('conference_info');

    const date = metadata.get('date');
    const publicationInfo = metadata.get('publication_info');
    const eprints = metadata.get('arxiv_eprints');
    const collaborations = metadata.get('collaborations');
    const collaborationsWithSuffix = metadata.get('collaborations_with_suffix');
    const canEdit = metadata.get('can_edit', false);

    return (
      <ResultItem
        leftActions={
          <Fragment>
            {fullTextLinks && (
              <UrlsAction
                urls={fullTextLinks}
                icon={<FilePdfOutlined />}
                iconText="pdf"
                trackerEventId="PdfDownload"
              />
            )}
            {dois && <DOILinkAction dois={dois} />}
            <CiteModalActionContainer recordId={recordId} />
            {canEdit && (
              <EditRecordAction pidType="literature" pidValue={recordId} />
            )}
          </Fragment>
        }
        rightActions={
          <Fragment>
            {citationCount != null && (
              <ListItemAction>
                <EventTracker eventId="Citations:Search">
                  <Link to={`${LITERATURE}?q=refersto:recid:${recordId}`}>
                    <IconText
                      text={`${citationCount} ${pluralizeUnlessSingle(
                        'citation',
                        citationCount
                      )}`}
                      icon={<LoginOutlined />}
                    />
                  </Link>
                </EventTracker>
              </ListItemAction>
            )}
          </Fragment>
        }
      >
        <div className="flex flex-nowrap">
          <div className="flex-grow-1">
            <Link className="f5" to={`${LITERATURE}/${recordId}`}>
              <LiteratureTitle title={title} />
            </Link>
          </div>
          <ResponsiveView
            min="sm"
            render={() => <div className="light-silver pl2">#{searchRank}</div>}
          />
        </div>
        <div className="mt1">
          <AuthorsAndCollaborations
            authorCount={authorCount}
            authors={authors}
            collaborations={collaborations}
            collaborationsWithSuffix={collaborationsWithSuffix}
          />
          {date && (
            <>
              {' ('}
              <LiteratureDate date={date} />
              {')'}
            </>
          )}
        </div>
        <div className="mt1">
          <InlineUL separator={SEPARATOR_MIDDLEDOT}>
            {publicationInfo && (
              <PublicationInfoList publicationInfo={publicationInfo} />
            )}
            {conferenceInfo && (
              <ConferenceInfoList conferenceInfo={conferenceInfo} />
            )}
            {eprints && <ArxivEprintList eprints={eprints} />}
          </InlineUL>
        </div>
      </ResultItem>
    );
  }
}

LiteratureItem.propTypes = {
  metadata: PropTypes.instanceOf(Map).isRequired,
  searchRank: PropTypes.number.isRequired,
};

export default LiteratureItem;
