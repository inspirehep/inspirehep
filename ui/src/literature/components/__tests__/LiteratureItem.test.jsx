import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import LiteratureItem from '../LiteratureItem';

describe('LiteratureItem', () => {
  it('renders with all props set, including sub props', () => {
    const metadata = fromJS({
      date: 'Jun 7, 1993',
      titles: [{ title: 'test' }],
      authors: [{ full_name: 'Test, Author' }],
      arxiv_eprints: [{ value: '1234567890' }],
      fulltext_links: [{ value: 'https://fulltext.cern/pdf/1' }],
      urls: [{ value: 'http://lss.fnal.gov/conf/C8206282/pg209.pdf' }],
      control_number: 12345,
      citation_count: 12435,
      publication_info: [{ journal_title: 'Test Jornal' }],
      collaborations: [{ value: 'CMS' }],
      collaborations_with_suffix: [{ value: 'CMS Group' }],
      conference_info: [
        {
          acronyms: ['MG15', 'SAP16'],
          titles: [
            {
              title:
                '15th Marcel Grossmann Meeting on Recent Developments in Theoretical and Experimental General Relativity, Astrophysics, and Relativistic Field Theories',
            },
          ],
        },
      ],
    });
    const wrapper = shallow(
      <LiteratureItem metadata={metadata} searchRank={2} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('does not arxiv pdf download action if there is no eprint value', () => {
    const metadata = fromJS({
      control_number: 12345,
      titles: [{ title: 'test' }],
    });
    const wrapper = shallow(
      <LiteratureItem metadata={metadata} searchRank={1} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders 0 citations if it does not exist', () => {
    const metadata = fromJS({
      control_number: 12345,
      titles: [{ title: 'test' }],
    });
    const wrapper = shallow(
      <LiteratureItem metadata={metadata} searchRank={2} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders publication info when book exists but publication title not', () => {
    const metadata = fromJS({
      control_number: 12345,
      titles: [{ title: 'test' }],
      publication_info: [
        {
          year: 2021,
          journal_volume: '9783030795993',
        },
      ],
      document_type: ['book'],
      book_series: [
        {
          title: 'SpringerBriefs in Physics',
        },
      ],
    });
    const wrapper = shallow(
      <LiteratureItem metadata={metadata} searchRank={3} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
