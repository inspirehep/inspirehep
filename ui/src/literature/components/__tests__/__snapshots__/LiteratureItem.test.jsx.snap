// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteratureItem does not arxiv pdf download action if there is no eprint value 1`] = `
<ResultItem
  leftActions={
    <React.Fragment>
      <Connect(CiteModalAction)
        recordId={12345}
      />
    </React.Fragment>
  }
  rightActions={
    <React.Fragment>
      <ListItemAction>
        <EventTracker
          eventId="Citations:Search"
          eventPropName="onClick"
          extractEventArgsToForward={null}
        >
          <ForwardRef
            to="/literature?q=refersto:recid:12345"
          >
            <IconText
              icon={<ForwardRef(LoginOutlined) />}
              text="0 citations"
            />
          </ForwardRef>
        </EventTracker>
      </ListItemAction>
    </React.Fragment>
  }
>
  <div
    className="flex flex-nowrap"
  >
    <div
      className="flex-grow-1"
    >
      <Link
        className="f5"
        to="/literature/12345"
      >
        <LiteratureTitle
          title={
            Immutable.Map {
              "title": "test",
            }
          }
        />
      </Link>
    </div>
    <ResponsiveView
      min="sm"
      render={[Function]}
    />
  </div>
  <div
    className="mt1"
  >
    <AuthorsAndCollaborations
      authors={Immutable.List []}
      collaborations={Immutable.List []}
      collaborationsWithSuffix={Immutable.List []}
      enableAuthorsShowAll={false}
    />
  </div>
  <div
    className="mt1"
  >
    <InlineUL
      separator=" • "
      wrapperClassName={null}
    />
  </div>
</ResultItem>
`;

exports[`LiteratureItem renders 0 citations if it does not exist 1`] = `
<ResultItem
  leftActions={
    <React.Fragment>
      <Connect(CiteModalAction)
        recordId={12345}
      />
    </React.Fragment>
  }
  rightActions={
    <React.Fragment>
      <ListItemAction>
        <EventTracker
          eventId="Citations:Search"
          eventPropName="onClick"
          extractEventArgsToForward={null}
        >
          <ForwardRef
            to="/literature?q=refersto:recid:12345"
          >
            <IconText
              icon={<ForwardRef(LoginOutlined) />}
              text="0 citations"
            />
          </ForwardRef>
        </EventTracker>
      </ListItemAction>
    </React.Fragment>
  }
>
  <div
    className="flex flex-nowrap"
  >
    <div
      className="flex-grow-1"
    >
      <Link
        className="f5"
        to="/literature/12345"
      >
        <LiteratureTitle
          title={
            Immutable.Map {
              "title": "test",
            }
          }
        />
      </Link>
    </div>
    <ResponsiveView
      min="sm"
      render={[Function]}
    />
  </div>
  <div
    className="mt1"
  >
    <AuthorsAndCollaborations
      authors={Immutable.List []}
      collaborations={Immutable.List []}
      collaborationsWithSuffix={Immutable.List []}
      enableAuthorsShowAll={false}
    />
  </div>
  <div
    className="mt1"
  >
    <InlineUL
      separator=" • "
      wrapperClassName={null}
    />
  </div>
</ResultItem>
`;

exports[`LiteratureItem renders with all props set, including sub props 1`] = `
<ResultItem
  leftActions={
    <React.Fragment>
      <UrlsAction
        icon={<ForwardRef(FilePdfOutlined) />}
        iconText="pdf"
        trackerEventId="PdfDownload"
        urls={
          Immutable.List [
            Immutable.Map {
              "value": "https://fulltext.cern/pdf/1",
            },
          ]
        }
      />
      <Connect(CiteModalAction)
        recordId={12345}
      />
    </React.Fragment>
  }
  rightActions={
    <React.Fragment>
      <ListItemAction>
        <EventTracker
          eventId="Citations:Search"
          eventPropName="onClick"
          extractEventArgsToForward={null}
        >
          <ForwardRef
            to="/literature?q=refersto:recid:12345"
          >
            <IconText
              icon={<ForwardRef(LoginOutlined) />}
              text="1 citation"
            />
          </ForwardRef>
        </EventTracker>
      </ListItemAction>
    </React.Fragment>
  }
>
  <div
    className="flex flex-nowrap"
  >
    <div
      className="flex-grow-1"
    >
      <Link
        className="f5"
        to="/literature/12345"
      >
        <LiteratureTitle
          title={
            Immutable.Map {
              "title": "test",
            }
          }
        />
      </Link>
    </div>
    <ResponsiveView
      min="sm"
      render={[Function]}
    />
  </div>
  <div
    className="mt1"
  >
    <AuthorsAndCollaborations
      authors={
        Immutable.List [
          Immutable.Map {
            "full_name": "Test, Author",
          },
        ]
      }
      collaborations={
        Immutable.List [
          Immutable.Map {
            "value": "CMS",
          },
        ]
      }
      collaborationsWithSuffix={
        Immutable.List [
          Immutable.Map {
            "value": "CMS Group",
          },
        ]
      }
      enableAuthorsShowAll={false}
    />
     (
    <LiteratureDate
      date="Jun 7, 1993"
    />
    )
  </div>
  <div
    className="mt1"
  >
    <InlineUL
      separator=" • "
      wrapperClassName={null}
    >
      <PublicationInfoList
        labeled={true}
        publicationInfo={
          Immutable.List [
            Immutable.Map {
              "journal_title": "Test Jornal",
            },
          ]
        }
      />
      <ConferenceInfoList
        conferenceInfo={
          Immutable.List [
            Immutable.Map {
              "acronyms": Immutable.List [
                "MG15",
                "SAP16",
              ],
              "titles": Immutable.List [
                Immutable.Map {
                  "title": "15th Marcel Grossmann Meeting on Recent Developments in Theoretical and Experimental General Relativity, Astrophysics, and Relativistic Field Theories",
                },
              ],
            },
          ]
        }
        wrapperClassName={null}
      />
      <ArxivEprintList
        eprints={
          Immutable.List [
            Immutable.Map {
              "value": "1234567890",
            },
          ]
        }
      />
    </InlineUL>
  </div>
</ResultItem>
`;
